OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs

```diff
@@ -125,39 +125,69 @@ protected static bool HasAmmo(Actor a)
 
 		protected static bool ReloadsAutomatically(Actor a)
 		{
-			var ammoPools = a.TraitsImplementing<AmmoPool>();
 			var rearmable = a.TraitOrDefault<Rearmable>();
 			if (rearmable == null)
 				return true;
 
+			var ammoPools = a.TraitsImplementing<AmmoPool>();
 			return ammoPools.All(ap => !rearmable.Info.AmmoPools.Contains(ap.Info.Name));
 		}
 
-		protected static bool IsRearm(Actor a)
+		// Checks the number of anti air enemies around units
+		protected virtual bool ShouldFlee(Squad owner)
 		{
-			if (a.IsIdle)
-				return false;
+			return ShouldFlee(owner, enemies => CountAntiAirUnits(enemies) * MissileUnitMultiplier > owner.Units.Count);
+		}
 
-			var activity = a.CurrentActivity;
-			var type = activity.GetType();
-			if (type == typeof(Resupply))
-				return true;
+		// Retreat units from combat, or for supply only in idle
+		protected void Retreat(Squad owner, bool resupplyonly)
+		{
+			// Reload units.
+			foreach (var a in owner.Units)
+			{
+				if (!ReloadsAutomatically(a) && !FullAmmo(a))
+				{
+					if (IsRearm(a))
+						continue;
 
-			var next = activity.NextActivity;
-			if (next == null)
-				return false;
+					owner.Bot.QueueOrder(new Order("ReturnToBase", a, false));
+					continue;
+				}
+				else if (!resupplyonly)
+					owner.Bot.QueueOrder(new Order("Move", a, Target.FromCell(owner.World, RandomBuildingLocation(owner)), false));
+			}
 
-			var nextType = next.GetType();
-			if (nextType == typeof(Resupply))
-				return true;
+			// Repair units. One by one to avoid give out mass orders
+			foreach (var a in owner.Units)
+			{
+				if (IsRearm(a))
+					continue;
 
-			return false;
-		}
+				Actor repairBuilding = null;
+				var orderId = "Repair";
 
-		// Checks the number of anti air enemies around units
-		protected virtual bool ShouldFlee(Squad owner)
-		{
-			return ShouldFlee(owner, enemies => CountAntiAirUnits(enemies) * MissileUnitMultiplier > owner.Units.Count);
+				if (a.TraitOrDefault<IHealth>() != null && a.TraitOrDefault<IHealth>().DamageState > DamageState.Undamaged)
```

Comment:
You should cache the result of `TraitOrDefault` in a local variable instead of querying it twice.